File: py/constellation.py
#!/usr/bin/env python
# coding: utf-8

# In[291]:


import numpy as np
import json
from typing import NamedTuple


# In[292]:


class Parameters(object):
    pass

Const = Parameters()
Const.earthRadius = 6378135;           # Экваториальный радиус Земли [km]
Const.earthGM = 3.986004415e+14;       # Гравитационный параметр Земли [m3/s2]
Const.earthJ2 = 1.082626e-3;           # First zonal harmonic coefficient in the expansion of the Earth's gravity field

group = Parameters()


# In[293]:


class Walker(NamedTuple):
    inclination:  float
    satsPerPlane: int
    planeCount:   int 
    f:            int
    altitude:     float
    maxRaan:      float
    startRaan:    float
                
class WalkerGroup(Walker):

    def getTotalSatCount(self): 
        return self.satsPerPlane * self.planeCount;
    
    def getInitialElements(self):
        startRaan = np.deg2rad(self.startRaan)
        maxRaan = np.deg2rad(self.maxRaan)
        inclination = np.deg2rad(self.inclination)
        altitude = self.altitude * 1000
        
        raans = np.linspace(startRaan, startRaan + maxRaan, self.planeCount + 1)
        raans = raans[0:-1] % (2 * np.pi)

        elements = np.zeros((self.getTotalSatCount(), 6))
        idx = 0
        raanIdx = 0
        for raan in raans:
            for satIdx in range(self.satsPerPlane):
                sma = Const.earthRadius + altitude
                aol = 2 * np.pi / self.satsPerPlane * (satIdx + 1) + 2 * np.pi / self.getTotalSatCount() * self.f * raanIdx

                elements[idx, :] = [sma, 0, 0, raan, inclination, aol]
                idx += 1
                    
            raanIdx += 1
            
        return elements


# In[294]:


class Constellation:
    
    def __init__(self, nameCode):
        self.totalSatCount = 0
        self.groups = []
        self.elements = []
        self.loadFromConfig(nameCode)
        
    def loadFromConfig(self, nameCode):
        f = open('ConstellationsTest.json')
        jsonData = json.loads(f.read())
        
        for entryIdx in range(len(jsonData)):
            if (jsonData[entryIdx]['name']).lower() == nameCode.lower():
                print("Загружена группировка " + nameCode)
                constellationData = jsonData[entryIdx]
                
                for groupIdx in range(len(constellationData['Walkers'])):
                    self.groups.append(WalkerGroup(*constellationData['Walkers'][groupIdx]))
                    self.totalSatCount += self.groups[groupIdx].getTotalSatCount()
                
                f.close()
                return 
            
        f.close()
        raise Exception('Группировка не найдена в файле')
        
    def getInitialState(self):
        self.elements = np.zeros((self.totalSatCount, 6));
        shift = 0;

        for group in self.groups:
            ending = shift + group.getTotalSatCount()
            self.elements[shift:ending, :] = group.getInitialElements()
            shift = ending
            

    def propagateJ2(self, epochs):
        self.stateEci = np.zeros((self.totalSatCount, 3, len(epochs)))

        inclination = self.elements[:, 4]
        sma         = self.elements[:, 0]
        Omega0      = np.sqrt(Const.earthGM / sma**3)
        aol0        = self.elements[:, 5]

        raanPrecessionRate = -1.5 * (Const.earthJ2 * np.sqrt(Const.earthGM) * Const.earthRadius**2) / (sma**(7/2)) * np.cos(inclination)
        draconicOmega      = np.sqrt(Const.earthGM / sma**3) * (1 - 1.5 * Const.earthJ2 * (Const.earthRadius / sma)**2) * (1 - 4 * np.cos(inclination)**2)

        breakpoint
        
        for epoch in epochs:
            aol = aol0 + epoch * draconicOmega
            Omega = Omega0 + epoch * raanPrecessionRate

            epochState = sma * [(np.cos(aol) * np.cos(Omega) - np.sin(aol) * np.cos(inclination) * np.sin(Omega)), 
                                (np.cos(aol) * np.sin(Omega) + np.sin(aol) * np.cos(inclination) * np.cos(Omega)), 
                                (np.sin(aol) * np.sin(inclination))]
            self.stateEci[:, :, epochs.index(epoch)]  = np.array(epochState).T 


# In[ ]:

File: MATLAB/Constellation.m
classdef Constellation < handle

    properties
        totalSatCount = 0;
        groups = {};
        state;

        % константы
        earthRadius = 6378135;           % Экваториальный радиус Земли [m]
        earthGM = 3.986004415e+14;       % Гравитационный параметр Земли [m3/s2]
        earthJ2 = 1.082626e-3;           % Вторая зональная гармоника геопотенциала
    end

    methods

        function this = Constellation(varargin)
            if isempty(varargin)
                return
            end
            this.loadFromConfigFile(varargin{1});
        end

        function loadFromConfigFile(this, code)
            fileName = 'constellationsTest.json';
            str = fileread(fileName);
            data = jsondecode(str);
            dataThis = [];

            for i = 1:length(data)
                if strcmpi(data(i).name, code)
                    dataThis = data(i);
                    break
                end
            end

            if isempty(dataThis)
                disp('Группировка не найдена в файле');
                return
            end

            for i = 1:size(dataThis.Walkers, 1)
                group.inclination = deg2rad(dataThis.Walkers(i, 1));        % наклонение орбитальной плоскости
                group.satsPerPlane = dataThis.Walkers(i, 2);				% число КА в каждой орбитальной плоскости группы
                group.planeCount = dataThis.Walkers(i, 3);					% число орбитальных плоскостей в группе
                group.f = dataThis.Walkers(i, 4);							% фазовый сдвиг по аргументу широты между КА в соседних плоскостях
                group.altitude = dataThis.Walkers(i, 5);					% высота орбиты
                group.maxRaan = deg2rad(dataThis.Walkers(i, 6));            % максимум прямого восхождения восходящего узла (при распределении орбитальных плоскостей)
                group.startRaan = deg2rad(dataThis.Walkers(i, 7));			% прямое восхождение восходящего узла для первой плоскости
                group.totalSatCount = group.satsPerPlane * group.planeCount;

                this.groups{length(this.groups) + 1} = group;                
                this.totalSatCount = this.totalSatCount + group.totalSatCount;
            end
        end

        function getInitialState(this)
            this.state.elements = zeros(this.totalSatCount, 6);
            shift = 1;

            for group = this.groups
                for i = 1:length(group{1})
                    ending = shift + group{1}(i).totalSatCount - 1;
                    this.state.elements(shift:ending,:) = this.getInitialElements(group{1});
                    shift = ending + 1;
                end
            end
        end

        function elements = getInitialElements(this, group)
            raans = linspace(group.startRaan, group.startRaan + group.maxRaan, group.planeCount + 1);
            raans = mod(raans(1:end-1), 2 * pi);

            elements = zeros(group.totalSatCount, 6);
            idx = 1;
            raanIDX = 0;
            for raan = raans
                for i = 0:group.satsPerPlane-1
                    sma = this.earthRadius + group.altitude * 1000;
                    aol = 2 * pi / group.satsPerPlane * i + 2 * pi / group.totalSatCount * group.f * raanIDX;

                    elements(idx, :) = [sma, 0, 0, raan, group.inclination, aol];
                    idx = idx + 1;
                end
                raanIDX = raanIDX + 1;
            end
        end        

        function propagateJ2(this, epochs)
            this.state.eci = zeros(this.totalSatCount, 3, length(epochs));

            sma         = this.state.elements(:, 1);
            inclination = this.state.elements(:, 5);            
            raan0       = this.state.elements(:, 4);
            aol0        = this.state.elements(:, 6);

            raanPrecessionRate = -1.5 * (this.earthJ2 * this.earthGM^(1/2) * this.earthRadius^2) ./ (sma.^(7/2)) .* cos(inclination);
            draconicOmega      = sqrt(this.earthGM ./ sma.^3) .* (1 - 1.5 * this.earthJ2 .* (this.earthRadius ./ sma).^2) .* (1 - 4 .* cos(inclination).^2);

            for epochIdx = 1:length(epochs)
                aol = aol0 + epochs(epochIdx) * draconicOmega;
                raanOmega = raan0 + epochs(epochIdx) * raanPrecessionRate;

                this.state.eci(:, :, epochIdx)  = [sma .* (cos(aol) .* cos(raanOmega) - sin(aol) .* cos(inclination) .* sin(raanOmega)), ...
                                                   sma .* (cos(aol) .* sin(raanOmega) + sin(aol) .* cos(inclination) .* cos(raanOmega)), ...
                                                   sma .* (sin(aol) .* sin(inclination))];
            end
        end
    end
end
